[Llama] Refactor Llama

LittleHeroZZZX opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/PaddleFormers/pull/ -->
#### Before submitting

- [ x ] Lint code. If there are lint issues, please format the code first.

```shell
# Install and register `pre-commit` in the project folder
pip install pre-commit && pre-commit install

# Process previous code files separately
pre-commit run --file XXXX.py
```

- [ ] Add test cases into `tests` folder. If there are codecov issues, please add tests cases first.

### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
Others 
### PR changes
<!-- One of [ Models | APIs | Docs | Others ] -->
Models 
### Description
<!-- Describe what this PR does -->
Refactor Llama3

LittleHeroZZZX committed:
Refactor llama

LittleHeroZZZX commented on paddleformers/transformers/llama/Llama2.LICENSE:
许可证文件是否要保留？

LittleHeroZZZX committed:
Fix a lot

LittleHeroZZZX commented on tests/transformers/test_modeling_common.py:
```diff
@@ -607,13 +607,11 @@ def test_resize_tokens_embeddings(self):
             model(**self._prepare_for_class(inputs_dict, model_class))
 
             # Check that adding and removing tokens has not modified the first part of the embedding matrix.
-            models_equal = True
-            for p1, p2 in zip(cloned_embeddings, model_embed.weight):
-                if not paddle.equal_all(p1, p2).item():
-                    models_equal = False
-                    break
-
-            self.assertTrue(models_equal)
```
原版实现逐元素比非常非常慢，直接切片出对应元素来比

LittleHeroZZZX commented on paddleformers/transformers/conversion_utils.py:
```diff
@@ -1250,7 +1250,7 @@ def convert_transpose_selected_weights(state_dict: dict, transpose_weight_keys:
                 if "lora" in key or value.ndim != 2:
                     continue
                 for trans_key in transpose_weight_keys:
-                    if re.search(f"\.{trans_key}\.weight$", key) or re.fullmatch(f"^{trans_key}\.weight$", key):
```
消除 warning

LittleHeroZZZX committed:
Rerun

lugimzzz commented on paddleformers/transformers/llama/configuration.py:
```diff
@@ -12,191 +11,73 @@
         vocab_size=32000,
         hidden_size=4096,
         intermediate_size=11008,
         max_position_embeddings=2048,
```
没有用到的config配置删掉，对比一下和这个文件key的差别，多余的都删除
https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/configuration_llama.py

lugimzzz replied:
如有需要看是否已经写在通用类中，如果也没有需要单独讨论

LittleHeroZZZX replied:
done

lugimzzz commented on paddleformers/transformers/llama/configuration.py:
```diff
@@ -12,191 +11,73 @@
-        fuse_attention_qkv=False,
-        fuse_attention_ffn=False,
+        use_flash_attention=False,
+        recompute=False,
```
recompute相关都写在基类了可以删掉，use_flash_attention也一样
https://github.com/PaddlePaddle/PaddleFormers/blob/develop/paddleformers/transformers/configuration_utils.py#L267

LittleHeroZZZX replied:
done

lugimzzz commented on paddleformers/transformers/llama/configuration.py:
```diff
@@ -12,191 +11,73 @@
-        immediate_clear_past_key_value=False,
-        dpo_config=None,
+        rope_theta=10000.0,
+        fuse_linear=False,
```
fuse_linear也一样

LittleHeroZZZX replied:
done

lugimzzz commented on paddleformers/transformers/llama/configuration.py:
```diff
@@ -12,191 +11,73 @@
+            self.rope_scaling["rope_type"] = self.rope_scaling["type"]
 
         super().__init__(
             pad_token_id=pad_token_id,
```
关注一下init里除了 standardize_rope_params&rope_config_validation 无法和transformers对齐原因https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/configuration_llama.py

LittleHeroZZZX replied:
修改后有三行不同：
```python
  self.rope_theta = kwargs.get("rope_theta", 10000.0)
  if self.rope_scaling is not None and "type" in self.rope_scaling:
      self.rope_scaling["rope_type"] = self.rope_scaling["type"]
```
第一行来自 standardize_rope_params&rope_config_validation，后两行借鉴自 https://github.com/PaddlePaddle/PaddleFormers/blob/38c47d4fe031ac924eea364f3c809a9d3a758f56/paddleformers/transformers/qwen3/configuration.py#L184-L186

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -12,1355 +11,311 @@
-        # Paddle Flash Attention input [ bz, seqlen, nhead, head_dim]
-        # Torch Flash Attention input [ bz, nhead, seqlen, head_dim]
+KvCacheItem = list[paddle.Tensor] | None
+KvCache = tuple[KvCacheItem, ...]
```
KvCacheItem作用？

LittleHeroZZZX replied:
用来做类型标注的别名，已替换成原本的类型

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -12,1355 +11,311 @@
-    return mask
+from .configuration import LlamaConfig
 
+token_id = 0
```
token_id作用？

LittleHeroZZZX replied:
已清理，调试时遗留下来的

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -12,1355 +11,311 @@
-        self.recompute_granularity = config.recompute_granularity
+        self.self_attn = LLamaAttention(config=config, layer_idx=layer_idx)
+        self.mlp = MLP(config)
+        self.input_layernorm = GeneralNorm.create(
```
https://github.com/PaddlePaddle/PaddleFormers/blob/develop/paddleformers/transformers/qwen2/modeling.py#L217
需要标注input_is_parallel=config.sequence_parallel
所有的layernorm都要注意这个问题

LittleHeroZZZX replied:
Done

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -12,1355 +11,311 @@
-class LlamaPretrainedModel(PretrainedModel):
-    config_class = LlamaConfig
-    base_model_prefix = "llama"
+class LlamaRotaryEmbedding(nn.Layer):
```
https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L128 LlamaRotaryEmbedding部分是否能够与transformers实现逐位对齐且训练过程中，保持float32的计算精度

LittleHeroZZZX replied:
Done. 组网代码与 transformers 重新做了对齐，单卡推理前平均 diff 从 0.0312 降到 0.0302

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -12,1355 +11,311 @@
-            outputs = outputs[0]
-
-        return outputs
+        return attn_output, attn_weights, past_key_value
```
https://github.com/PaddlePaddle/PaddleFormers/blob/develop/paddleformers/transformers/qwen2/modeling.py#L80 处理参考qwen，不需要attn_weights

lugimzzz replied:
不要 KvCacheItem

LittleHeroZZZX replied:
> https://github.com/PaddlePaddle/PaddleFormers/blob/develop/paddleformers/transformers/qwen2/modeling.py#L80 处理参考qwen，不需要attn_weights

完全移除 output_attentions 这个参数？

LittleHeroZZZX replied:
> 完全移除 output_attentions 这个参数？

Done

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -12,1355 +11,311 @@
+        value_states = self.v_proj(hidden_states).view(kv_shape)
 
-                query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
+        if position_embeddings is None:
```
为啥这里要新增一个raise error？如果fuse rope可以不传position_embedding的

LittleHeroZZZX replied:
已移除，原本不清楚这个机制。

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -1370,789 +325,328 @@ def _get_tensor_parallel_mappings(cls, config: LlamaConfig, is_split=True):
-            attn_mask_startend_row_indices,
-            use_reentrant=self.config.recompute_use_reentrant,
+        self.norm = GeneralNorm.create(
+            config=config,
```
norm相关问题同上

LittleHeroZZZX replied:
Done.

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -12,1355 +11,311 @@
+            norm_eps=config.rms_norm_eps,
+        )
+
+        if self.config.sequence_parallel:
```
if self.config.sequence_parallel:
            self.post_attention_layernorm.enable_sequence_parallel()
            if not hasattr(self.config, "disable_ffn_model_parallel"):
                self.input_layernorm.enable_sequence_parallel() 这个可以去掉

LittleHeroZZZX replied:
Done

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -12,1355 +11,311 @@
+    ) -> (
+        tuple[paddle.Tensor]
+        | tuple[paddle.Tensor, paddle.Tensor]
+        | tuple[paddle.Tensor, KvCacheItem]
```
不要KvCacheItem

LittleHeroZZZX replied:
Done

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -12,1355 +11,311 @@
-        if use_cache:
-            present_key_value = outputs[2 if output_attentions else 1]
-
+        hidden_states, attn_weights, current_key_value = self.self_attn(
```
去掉attn_weights

LittleHeroZZZX replied:
Done

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -12,1355 +11,311 @@
+            past_key_value=past_key_value,
+            use_cache=use_cache,
+        )
+        if TYPE_CHECKING:
```
不要if TYPE_CHECKING，参考qwen写法https://github.com/PaddlePaddle/PaddleFormers/blob/develop/paddleformers/transformers/qwen2/modeling.py#L247

LittleHeroZZZX replied:
Done

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -12,1355 +11,311 @@
+        "gate_proj",
+        "up_proj",
+        "down_proj",
+        "gate",
```
哪来的gate？"lm_head"不用了

LittleHeroZZZX replied:
直接复制的其它模型的 gate，已清理。

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -1370,789 +325,328 @@ def _get_tensor_parallel_mappings(cls, config: LlamaConfig, is_split=True):
-                    for i in range(num_layers):
-                        final_actions[key.replace("layers.0.", f"layers.{i}.")] = action
-                final_actions[key] = action
+        # lm_head.bias is always False
```
注释掉的代码删掉

LittleHeroZZZX replied:
Done

lugimzzz commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -1370,789 +325,328 @@ def _get_tensor_parallel_mappings(cls, config: LlamaConfig, is_split=True):
+        return_dict: bool = False,  # true when decode, false when pretrain & eval
+        **kwargs,
     ):
+        global token_id
```
这个东西的作用？

LittleHeroZZZX replied:
调试用，已清理。

lugimzzz commented on tests/transformers/llama/test_modeling.py:
```diff
@@ -394,7 +394,7 @@ def test_inference_with_attention(self):
+    internal_testing_model = "test_paddleformers/tiny-random-llama"
 
 
 # class LlamaCompatibilityTest(unittest.TestCase):
```
不要关掉注释跑 https://github.com/PaddlePaddle/PaddleFormers/blob/develop/tests/transformers/ernie4_5/test_modeling.py#L444C7-L444C32

LittleHeroZZZX committed:
Refine on comment and add llama3 intergration test

LittleHeroZZZX committed:
Fix unittest: update expected output values based on paddle-1104 and use llama3-shard in test_shard_checkpoint

LittleHeroZZZX committed:
Fix unittest: set config.head_dim after modify config.hidden_size

LittleHeroZZZX committed:
Fix unittest: set config.head_dim after modify config.hidden_size

LittleHeroZZZX committed:
Merge remote-tracking branch 'origin/develop' into llama3

LittleHeroZZZX committed:
Fix lm_head.weight tp mapping and miss tokenizer.pad_token_id

LittleHeroZZZX committed:
Supoort llama3 rope_type

LittleHeroZZZX committed:
Merge remote-tracking branch 'origin/develop' into llama3

LittleHeroZZZX committed:
Migrate fc

LittleHeroZZZX committed:
Merge develop , fix merge kit

LittleHeroZZZX committed:
rerun

LittleHeroZZZX committed:
rerun

WYB27 commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -12,1355 +11,338 @@
-        out = self.down_proj(x)
-        return out
-
+            attn_output = attn_output.reshape([batch_size, seq_len, -1]).contiguous()
```
这里需要对attn_output进行reshape+contiguous的原因是？attention_interface出来的维度不是这个维度吗

LittleHeroZZZX replied:
冗余操作，已移除。

LittleHeroZZZX committed:
Merge remote-tracking branch 'origin/develop' into llama3

LittleHeroZZZX committed:
Fix: remove uncessay reshape after attn_interface

WYB27 commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -1370,789 +350,373 @@ def _get_tensor_parallel_mappings(cls, config: LlamaConfig, is_split=True):
-                "The attn_mask_startend_row_indices will be used."
-            )
-            attention_mask = None
+        if attention_mask is not None and attention_mask.dtype != paddle.bool:
```
attention_mask和attn_mask_startend_row_indices有一个就可以，优先attn_mask_startend_row_indices。参考https://github.com/PaddlePaddle/PaddleFormers/blob/95943898637ed5442acdb68afc8613151192dfdc/paddleformers/transformers/glm4_moe/modeling.py#L1523

LittleHeroZZZX replied:
Done

WYB27 commented on paddleformers/transformers/llama/modeling.py:
```diff
@@ -1370,789 +350,373 @@ def _get_tensor_parallel_mappings(cls, config: LlamaConfig, is_split=True):
-            past_key_value = past_key_values[idx] if past_key_values is not None else None
+        # TODO(littleherozzzx): check self.config.fuse_rope
+        position_embeddings = self.rotary_emb(inputs_embeds, position_ids)
+        causal_mask = self._prepare_decoder_attention_mask(
```
causal_mask, attn_mask_startend_row_indices的处理方式，请参考通用生成方式
https://github.com/PaddlePaddle/PaddleFormers/blob/95943898637ed5442acdb68afc8613151192dfdc/paddleformers/transformers/glm4_moe/modeling.py#L1349C9-L1360C20

LittleHeroZZZX replied:
Done

WYB27 commented on paddleformers/transformers/model_outputs.py:
```diff
@@ -620,7 +620,7 @@ class BaseModelOutputWithPast(ModelOutput):
 
     last_hidden_state: paddle.Tensor = None
-    past_key_values: Optional[Tuple[Tuple[paddle.Tensor]]] = None
+    past_key_values: Optional[Tuple[Tuple[paddle.Tensor, paddle.Tensor]]] = None
```
这里从单Tensor改成双Tensor的原因？会影响其他模型吗

LittleHeroZZZX replied:
表示 past_key_values 这个 tuple 中的每个元素都是一个长度为 2 并且类型为 Tensor 的 tuple。原来表示写法表示长度 为 1，应该是与预期不符合的？

修改类型标注对运行时没有影响。

LittleHeroZZZX committed:
Empty commit to rerun ci

LittleHeroZZZX committed:
Refine on comment: use create_causal_masks_and_row_indices and warn if both attn_mask and attn_mask_start_..._ are provided

LittleHeroZZZX committed:
feat(llama): add comprehensive AOA configuration methods

Add `_gen_aoa_config` and `_gen_inv_aoa_config` class methods to LlamaPretrainedModel for generating detailed AOA (Array of Arrays) configuration mappings. These methods provide complete weight mapping specifications for all model components including embeddings, layer norms, attention projections, and MLP layers. The configuration handles both tied and untied word embeddings and supports layer-wise parameter substitution using $LAYER_ID placeholders. This enables proper weight transformation and mapping for model conversion and deployment scenarios.

LittleHeroZZZX committed:
Merge remote-tracking branch 'origin/develop' into llama3

LittleHeroZZZX committed:
feat(tests): add HF conversion test cases to shard checkpoint test

Add test cases for both converted and non-converted HuggingFace models in the shard checkpoint test. The test now iterates through model names with corresponding convert_from_hf flags to ensure both conversion paths work correctly with different low_cpu_mem_usage settings. This improves test coverage for model loading from pretrained with HuggingFace conversion.

WYB27 reviewed (approved):
LGTM
